nomad/job_endpoint_hooks.go

```diff
@@ -40,6 +40,8 @@ type jobValidator interface {
 }
 
 func (j *Job) admissionControllers(job *structs.Job) (out *structs.Job, warnings []error, err error) {
+	// Mutators run first before validators, so validators view the final rendered job.
+	// So, mutators must handle invalid jobs.
```

Comment:
I'm not fully sure what the motivation of having mutators run first: but this has bitten us before with a similar panic https://github.com/hashicorp/nomad/pull/6575 .

I'm not fully sure of the impact of re-ordering steps, so we validate first then mutate.  Skimming code, that would mean that we'd attempt to canonize the job first before validating it, which may open another can of worms.

Reply:
This is really unfortunate in that we really have two types of validation: validation for safety which really wants to happen before mutation (canonicalization, making sure we don't have nil pointers in unexpected places), and validation for "business logic" (ex. task group count for system jobs).

But solving that is beyond the scope of this bug fix, so :+1: for leaving this comment.